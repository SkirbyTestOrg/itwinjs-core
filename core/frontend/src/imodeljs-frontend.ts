/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

export * from "./tools/AccuDrawTool";
export * from "./tools/EditManipulator";
export * from "./tools/ElementSetTool";
export * from "./tools/EventController";
export * from "./tools/MeasureTool";
export * from "./tools/SelectTool";
export * from "./tools/ToolAdmin";
export * from "./tools/Tool";
export * from "./tools/ViewTool";
export * from "./tools/ClipViewTool";
export * from "./tools/PrimitiveTool";
export * from "./tools/IdleTool";
export * from "./tools/ToolAssistance";
export * from "./AccuDraw";
export * from "./AccuSnap";
export * from "./AuxCoordSys";
export * from "./DevTools";
export * from "./FrontendRequestContext";
export * from "./CategorySelectorState";
export * from "./ContextRealityModelState";
export * from "./SpatialClassification";
export * from "./DisplayStyleState";
export * from "./ElementLocateManager";
export * from "./EmphasizeElements";
export * from "./EntityState";
export * from "./FuzzySearch";
export * from "./GeoServices";
export * from "./HitDetail";
export * from "./IModelConnection";
export * from "./ImageUtil";
export * from "./FrontendLoggerCategory";
export * from "./Marker";
export * from "./ModelSelectorState";
export * from "./ModelState";
export * from "./NotificationManager";
export * from "./plugin/Plugin";
export * from "./plugin/PluginUiManager";
export * from "./RelativePosition";
export * from "./RenderScheduleState";
export * from "./SelectionSet";
export * from "./Sheet";
export * from "./Sprites";
export * from "./StandardView";
export * from "./SubCategoriesCache";
export * from "./TentativePoint";
export * from "./QuantityFormatter";
export * from "./ViewContext";
export * from "./ViewManager";
export * from "./Viewport";
export * from "./ViewState";
export * from "./IModelApp";
export * from "./NoRenderApp";
export * from "./tile/TileAdmin";
export * from "./tile/Tile";
export * from "./tile/TileTree";
export * from "./tile/WebMapTileTree";
export * from "./tile/BackgroundTerrainTileTree";
export * from "./tile/CesiumWorldTerrainTileTree";
export * from "./tile/BingElevation";
export * from "./render/FeatureSymbology";
export * from "./render/GraphicBuilder";
export * from "./render/MockRender";
export * from "./render/System";
export * from "./render/webgl/Target";
export * from "./RenderCompatibility";
export * from "./oidc/OidcBrowserClient";
export * from "./TerrainProvider";
export * from "./properties/Description";
export * from "./properties/EditorParams";
export * from "./properties/Record";
export * from "./properties/Value";
export * from "./properties/ToolSettingsValue";
export * from "./properties/PrimitiveTypes";
export * from "./properties/BaseQuantityDescription";
export * from "./properties/AngleDescription";
export * from "./properties/LengthDescription";
export * from "./SolarCalculate";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
declare var BUILD_TYPE: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("imodeljs-frontend", BUILD_SEMVER);
  (window as any).iModelJsVersions.set("buildType", BUILD_TYPE);
}

/** @docs-package-description
 * The ($frontend) package always runs in a web browser. It contains classes for [querying iModels and showing views]($docs/learning/frontend/index.md).
 */

/**
 * @docs-group-description IModelApp
 * Classes for configuring and administering an iModel.js application.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description IModelConnection
 * Classes for working with a connection to an [iModel briefcase]($docs/learning/IModels.md)
 */
/**
 * @docs-group-description ElementState
 * Classes for working with the *state* of Elements in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description ModelState
 * Classes for working with the *state* of Models in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Tools
 * Classes for [working with Tools]($docs/learning/frontend/Tools.md)
 */
/**
 * @docs-group-description Views
 * Classes for [working with Views]($docs/learning/frontend/Views.md)
 */
/**
 * @docs-group-description LocatingElements
 * Classes for locating and snapping to elements in views.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description AccuDraw
 * AccuDraw provides helpful assistance for creating and modifying elements in a view.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Notifications
 * Notifications provide feedback to the user of something of interest.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Plugins
 * Classes for creating and managing runtime [Plugins]($docs/learning/frontend/Plugins.md)
 */
/**
 * @docs-group-description Properties
 * Classes for working with property records and descriptions.
 */
/**
 * @docs-group-description Rendering
 * Classes for rendering the contents of views.
 */
/**
 * @docs-group-description SelectionSet
 * Classes for working with the set of selected elements.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
