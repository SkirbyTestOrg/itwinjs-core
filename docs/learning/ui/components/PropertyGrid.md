# PropertyGrid

## API Reference

* [PropertyGrid]($ui-components:PropertyGrid)
