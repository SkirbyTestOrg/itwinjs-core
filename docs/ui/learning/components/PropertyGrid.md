# Property Grid

## API Reference

* [PropertyGrid]($components:PropertyGrid)
