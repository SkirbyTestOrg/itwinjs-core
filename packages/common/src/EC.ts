/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2017 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import { Id64 } from "@bentley/bentleyjs-core";

export type ClassId = Id64;

export type InstanceId = Id64;

export interface InstanceKey {
  className: string;
  id: InstanceId;
}

export interface InstanceKeyJSON {
  className: string;
  id: string;
}

export const instanceKeyFromJSON = (json: InstanceKeyJSON): InstanceKey => {
  return { ...json, id: new Id64(json.id) };
};

export type InstanceKeysList = InstanceKey[];

/** Information about an ECClass. */
export interface ClassInfo {
  id: ClassId;
  name: string;
  label: string;
}

export interface EnumerationChoice {
  label: string;
  value: string | number;
}

export interface EnumerationInfo {
  choices: EnumerationChoice[];
  isStrict: boolean;
}

export interface KindOfQuantityInfo {
  name: string;
  label: string;
  persistenceUnit: string;
  currentFusId: string;
}

export interface PropertyInfo {
  classInfo: ClassInfo;
  name: string;
  type: string;
  enumerationInfo?: EnumerationInfo;
  kindOfQuantity?: KindOfQuantityInfo;
}

/** A structure that describes a related class and the properties of that relationship. */
export interface RelatedClassInfo {
  /** Information about the source ECClass */
  sourceClassInfo: ClassInfo;

  /** Information about the target ECClass */
  targetClassInfo: ClassInfo;

  /** Information about the relationship ECClass */
  relationshipInfo: ClassInfo;

  /** Should the relationship be followed in a forward direction to access the related class. */
  isForwardRelationship: boolean;

  /** Is the relationship handled polymorphically */
  isPolymorphicRelationship: boolean;
}

/** A structure that describes a related class path. */
export type RelationshipPathInfo = RelatedClassInfo[];
