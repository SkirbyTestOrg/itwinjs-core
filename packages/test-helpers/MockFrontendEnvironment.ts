/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
(global as any).WebGLRenderingContext = require("gl"); // tslint:disable-line:no-var-requires
(global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // tslint:disable-line:no-var-requires
