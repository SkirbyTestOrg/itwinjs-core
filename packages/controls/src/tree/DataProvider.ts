/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2017 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Hierarchies */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Node, NodeKey, PageOptions } from "@bentley/ecpresentation-common";
import { ECPresentation } from "@bentley/ecpresentation-frontend";
import { CheckBoxState } from "@bentley/ui-core";
import { TreeNodeItem } from "@bentley/ui-components";
import StyleHelper from "../common/StyleHelper";

class MemoizationHelpers {
  public static createKeyForPageOptions(pageOptions?: PageOptions) {
    if (!pageOptions)
      return "0/0";
    return `${(pageOptions.pageStart) ? pageOptions.pageStart : 0}/${(pageOptions.pageSize) ? pageOptions.pageSize : 0}`;
  }
  public static createKeyForTreeNodeItem(item: TreeNodeItem) { return item.id; }
  public static getRootNodesKeyResolver(pageOptions?: PageOptions) { return MemoizationHelpers.createKeyForPageOptions(pageOptions); }
  public static getChildNodesKeyResolver(parent: TreeNodeItem, pageOptions?: PageOptions) {
    return `${MemoizationHelpers.createKeyForTreeNodeItem(parent)}/${MemoizationHelpers.createKeyForPageOptions(pageOptions)}`;
  }
  public static getChildNodesCountKeyResolver(item: TreeNodeItem) { return MemoizationHelpers.createKeyForTreeNodeItem(item); }
}

/**
 * Presentation Rules-driven tree data provider.
 */
export default class TreeDataProvider {
  private _rulesetId: string;
  private _imodelConnection: IModelConnection;

  /**
   * Constructor.
   * @param imodelToken Token of the imodel to pull data from.
   * @param rulesetId Id of the ruleset used by this data provider.
   */
  public constructor(connection: IModelConnection, rulesetId: string) {
    this._rulesetId = rulesetId;
    this._imodelConnection = connection;
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetId; }
  public set rulesetId(value: string) {
    if (this._rulesetId === value)
      return;
    this._rulesetId = value;
    this.clearCaches();
  }

  /** [[IModelConnection]] used by this data provider */
  public get connection(): IModelConnection { return this._imodelConnection; }
  public set connection(value: IModelConnection) {
    if (this._imodelConnection === value)
      return;
    this._imodelConnection = value;
    this.clearCaches();
  }

  private clearCaches(): void {
    this.getRootNodesCount.cache.clear();
    this.getRootNodes.cache.clear();
    this.getChildNodesCount.cache.clear();
    this.getChildNodes.cache.clear();
  }

  /** Called to get extended options for node requests */
  private createRequestOptions(): object {
    return {
      RulesetId: this._rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return node.extendedData.key as NodeKey;
  }

  /**
   * Returns the root nodes.
   * @param pageOptions Information about the requested page of data.
   */
  public getRootNodes = _.memoize(async (pageOptions?: PageOptions): Promise<ReadonlyArray<Readonly<TreeNodeItem>>> => {
    const nodes = await ECPresentation.presentation.getRootNodes(this.connection, pageOptions, this.createRequestOptions());
    return createTreeNodeItems(nodes);
  }, MemoizationHelpers.getRootNodesKeyResolver);

  /**
   * Returns the total number of root nodes.
   */
  public getRootNodesCount = _.memoize(async (): Promise<number> => {
    return await ECPresentation.presentation.getRootNodesCount(this.connection, this.createRequestOptions());
  });

  /**
   * Returns child nodes.
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public getChildNodes = _.memoize(async (parentNode: TreeNodeItem, pageOptions?: PageOptions): Promise<ReadonlyArray<Readonly<TreeNodeItem>>> => {
    const parentKey = this.getNodeKey(parentNode);
    const nodes = await ECPresentation.presentation.getChildren(this.connection, parentKey, pageOptions, this.createRequestOptions());
    const items = createTreeNodeItems(nodes);
    items.forEach((item: TreeNodeItem) => {
      item.parentId = parentNode.id;
    });
    return items;
  }, MemoizationHelpers.getChildNodesKeyResolver);

  /**
   * Returns the total number of child nodes.
   * @param parentNode The parent node to return children count for.
   */
  public getChildNodesCount = _.memoize(async (parentNode: TreeNodeItem): Promise<number> => {
    const parentKey = this.getNodeKey(parentNode);
    return await ECPresentation.presentation.getChildrenCount(this.connection, parentKey, this.createRequestOptions());
  }, MemoizationHelpers.getChildNodesCountKeyResolver);
}

const createTreeNodeItems = (nodes: ReadonlyArray<Readonly<Node>>): TreeNodeItem[] => {
  const list = new Array<TreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node));
  return list;
};

const createTreeNodeItem = (node: Readonly<Node>): TreeNodeItem => {
  const item: TreeNodeItem = {
    id: [...node.key.pathFromRoot].reverse().join("/"),
    label: node.label,
    description: node.description || "",
    hasChildren: node.hasChildren || false,
    labelForeColor: StyleHelper.getForeColor(node),
    labelBackColor: StyleHelper.getBackColor(node),
    labelBold: StyleHelper.isBold(node),
    labelItalic: StyleHelper.isItalic(node),
    displayCheckBox: node.isCheckboxVisible || false,
    checkBoxState: node.isChecked ? CheckBoxState.On : CheckBoxState.Off,
    isCheckBoxEnabled: node.isCheckboxEnabled || false,
    extendedData: { key: node.key },
  };
  return item;
};
