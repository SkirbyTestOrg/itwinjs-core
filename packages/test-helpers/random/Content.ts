/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2017 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import * as faker from "faker";
import * as c from "../../common/src/content";
import { createRandomECClassInfo, createRandomRelationshipPath } from "./EC";
import { nullable } from "./Misc";

const createRandomSelectClassInfo = (): c.SelectClassInfo => {
  return {
    selectClassInfo: createRandomECClassInfo(),
    isSelectPolymorphic: faker.random.boolean(),
    pathToPrimaryClass: createRandomRelationshipPath(),
    relatedPropertyPaths: [createRandomRelationshipPath(1), createRandomRelationshipPath(1)],
  } as c.SelectClassInfo;
};

export const createRandomCategory = (): c.CategoryDescription => {
  return {
    name: faker.random.word(),
    label: faker.random.words(),
    description: faker.lorem.sentence(),
    priority: faker.random.number(),
    expand: faker.random.boolean(),
  } as c.CategoryDescription;
};

export const createRandomPrimitiveTypeDescription = (): c.TypeDescription => {
  return {
    valueFormat: c.PropertyValueFormat.Primitive,
    typeName: faker.database.type(),
  } as c.PrimitiveTypeDescription;
};

export const createRandomEditorDescription = (): c.EditorDescription => {
  return {
    name: faker.random.word(),
  } as c.EditorDescription;
};

const createRandomPrimitiveFieldJson = () => {
  return {
    category: createRandomCategory(),
    name: faker.random.word(),
    label: faker.random.words(),
    type: createRandomPrimitiveTypeDescription(),
    isReadonly: faker.random.boolean(),
    priority: faker.random.number(),
    editor: nullable(createRandomEditorDescription),
  };
};

export const createRandomPrimitiveField = (): c.Field => {
  return c.Field.fromJSON(createRandomPrimitiveFieldJson())!;
};

export const createRandomDescriptorJson = (displayType?: string) => {
  const selectClasses = [createRandomSelectClassInfo(), createRandomSelectClassInfo()];
  const fields = [createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson()];
  return {
    connectionId: faker.random.uuid(),
    inputKeysHash: faker.random.uuid(),
    contentOptions: faker.random.objectElement(),
    displayType: displayType || faker.lorem.words(),
    selectClasses,
    fields,
    contentFlags: 0,
  };
};

export const createRandomDescriptor = (displayType?: string): c.Descriptor => {
  return c.Descriptor.fromJSON(createRandomDescriptorJson(displayType))!;
};
