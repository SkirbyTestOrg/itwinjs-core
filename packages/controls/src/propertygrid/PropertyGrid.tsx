/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { KeySet } from "@bentley/ecpresentation-common";
import { ECPresentation, SelectionHandler, SelectionChangeEventArgs, ISelectionProvider } from "@bentley/ecpresentation-frontend";
import { Orientation } from "@bentley/ui-core";
import { PropertyGrid as BasePropertyGrid } from "@bentley/ui-components";
import DataProvider from "./DataProvider";

/**
 * Props for the [[PropertyGrid]] control.
 */
export interface Props {
  /** Orientation of properties display. */
  orientation: Orientation;

  /** iModel to pull data from */
  imodel: IModelConnection;

  /** Presentation ruleset to use for creating the content for the control */
  rulesetId: string;

  /** Optional ID for the control. The ID is also used as a selection handler name. */
  id?: string;

  /** Optional custom data provider implementation. */
  dataProvider?: DataProvider;

  /** @hidden */
  selectionHandler?: SelectionHandler;
}

/**
 * Presentation rules -driven property grid control which also participates in
 * unified selection.
 */
export default class PropertyGrid extends React.Component<Props> {

  private _dataProvider: DataProvider;
  private _selectionHandler: SelectionHandler;

  public constructor(props: Props, context?: any) {
    super(props, context);
    const key = props.id ? props.id : `PropertyGrid_${new Date().getTime()}`;
    this._selectionHandler = props.selectionHandler
      ? props.selectionHandler : new SelectionHandler(ECPresentation.selection, key, props.imodel, props.rulesetId);
    this._selectionHandler.onSelect = this.onSelectionChanged;
    this._dataProvider = props.dataProvider
      ? props.dataProvider : new DataProvider(props.imodel, props.rulesetId);
  }

  public componentWillUnmount() {
    this._selectionHandler.dispose();
  }

  public componentWillReceiveProps(nextProps: Props) {
    this._selectionHandler.imodel = nextProps.imodel;
    this._selectionHandler.rulesetId = nextProps.rulesetId;

    this._dataProvider.connection = nextProps.imodel;
    this._dataProvider.rulesetId = nextProps.rulesetId;
  }

  // tslint:disable-next-line:naming-convention
  private onSelectionChanged = (evt: SelectionChangeEventArgs, selectionProvider: ISelectionProvider): void => {
    for (let i = evt.level; i >= 0; i--) {
      const selection = selectionProvider.getSelection(this.props.imodel, i);
      const hasSelection = !selection.isEmpty;
      if (hasSelection) {
        this._dataProvider.keys = selection;
        return;
      }
    }
    this._dataProvider.keys = new KeySet();
  }

  /** Get selection handler used by the tree */
  public get selectionHandler(): SelectionHandler { return this._selectionHandler; }

  /** Get data provider of this tree */
  public get dataProvider(): DataProvider { return this._dataProvider; }

  public render() {
    return (
      <BasePropertyGrid
        orientation={this.props.orientation}
        dataProvider={this._dataProvider}
      />
    );
  }
}
