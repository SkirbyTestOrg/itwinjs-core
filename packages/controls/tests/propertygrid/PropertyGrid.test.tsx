/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import "@helpers/MockFrontendEnvironment";
import * as React from "react";
import { expect } from "chai";
import { mount, shallow } from "enzyme";
import * as moq from "@helpers/Mocks";
import * as faker from "faker";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import {
  ECPresentation, ECPresentationManager,
  SelectionHandler, SelectionManager, SelectionChangeEvent, SelectionChangeType, ISelectionProvider,
} from "@bentley/ecpresentation-frontend";
import { Orientation } from "@bentley/ui-core";
import { PropertyData, PropertyDataChangeEvent } from "@bentley/ui-components";
import DataProvider from "@src/propertygrid/DataProvider";
import PropertyGrid, { Props as PropertyGridProps } from "@src/propertygrid/PropertyGrid";
import { KeySet } from "@bentley/ecpresentation-common/lib";
import { createRandomECInstanceKey } from "@helpers/random/EC";

describe("PropertyGrid", () => {

  const imodelMock = moq.Mock.ofType<IModelConnection>();
  const dataProviderMock = moq.Mock.ofType<DataProvider>();
  const selectionHandlerMock = moq.Mock.ofType<SelectionHandler>();
  beforeEach(() => {
    selectionHandlerMock.reset();
    setupDataProvider();
  });

  const setupDataProvider = (propertyData?: PropertyData) => {
    dataProviderMock.reset();
    if (!propertyData) {
      propertyData = {
        label: faker.random.word(),
        description: faker.random.words(),
        categories: [],
        records: {},
      };
    }
    dataProviderMock.setup((x) => x.getData()).returns(async () => propertyData!);
    dataProviderMock.setup((x) => x.onDataChanged).returns(() => new PropertyDataChangeEvent());
  };

  it("mounts", () => {
    mount(<PropertyGrid
      orientation={Orientation.Horizontal}
      dataProvider={dataProviderMock.object}
      selectionHandler={selectionHandlerMock.object}
      imodel={imodelMock.object}
      rulesetId={faker.random.word()} />);
  });

  it("creates default implementation for selection handler and data provider when not provided through props", () => {
    const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
    selectionManagerMock.setup((x) => x.selectionChange).returns(() => new SelectionChangeEvent());
    ECPresentation.selection = selectionManagerMock.object;

    const presentationManagerMock = moq.Mock.ofType<ECPresentationManager>();
    presentationManagerMock
      .setup((x) => x.getContentDescriptor(imodelMock.object, moq.It.isAnyString(), moq.It.isAny(), undefined, moq.It.isAny()))
      .returns(async () => undefined);
    ECPresentation.presentation = presentationManagerMock.object;

    const rulesetId = faker.random.word();

    const tree = mount(<PropertyGrid
      orientation={Orientation.Vertical}
      imodel={imodelMock.object}
      rulesetId={rulesetId} />).instance() as PropertyGrid;

    expect(tree.selectionHandler.name).to.not.be.undefined;
    expect(tree.selectionHandler.rulesetId).to.eq(rulesetId);
    expect(tree.selectionHandler.imodel).to.eq(imodelMock.object);

    expect(tree.dataProvider.rulesetId).to.eq(rulesetId);
    expect(tree.dataProvider.connection).to.eq(imodelMock.object);
  });

  it("renders correctly", () => {
    expect(shallow(<PropertyGrid
      orientation={Orientation.Horizontal}
      id={faker.random.uuid()}
      dataProvider={dataProviderMock.object}
      selectionHandler={selectionHandlerMock.object}
      imodel={imodelMock.object}
      rulesetId={faker.random.word()} />)).to.matchSnapshot();
  });

  it("disposes selection handler when unmounts", () => {
    const tree = mount(<PropertyGrid
      orientation={Orientation.Vertical}
      dataProvider={dataProviderMock.object}
      selectionHandler={selectionHandlerMock.object}
      imodel={imodelMock.object}
      rulesetId={faker.random.word()} />);
    tree.unmount();
    selectionHandlerMock.verify((x) => x.dispose(), moq.Times.once());
  });

  it("updates selection handler and data provider when props change", () => {
    const tree = mount<PropertyGridProps>(<PropertyGrid
      orientation={Orientation.Horizontal}
      dataProvider={dataProviderMock.object}
      selectionHandler={selectionHandlerMock.object}
      imodel={imodelMock.object}
      rulesetId={faker.random.word()} />);

    const imodelMock2 = moq.Mock.ofType<IModelConnection>();
    const rulesetId2 = faker.random.word();

    tree.setProps({
      imodel: imodelMock2.object,
      rulesetId: rulesetId2,
    });

    selectionHandlerMock.verify((x) => x.imodel = imodelMock2.object, moq.Times.once());
    selectionHandlerMock.verify((x) => x.rulesetId = rulesetId2, moq.Times.once());

    dataProviderMock.verify((x) => x.connection = imodelMock2.object, moq.Times.once());
    dataProviderMock.verify((x) => x.rulesetId = rulesetId2, moq.Times.once());
  });

  describe("selection handling", () => {

    it("sets data provider keys to overall selection on selection changes", () => {
      const keysOverall = new KeySet([createRandomECInstanceKey(), createRandomECInstanceKey()]);
      const keysAdded = new KeySet([createRandomECInstanceKey()]);
      const selectionProviderMock = moq.Mock.ofType<ISelectionProvider>();
      selectionProviderMock.setup((x) => x.getSelection(imodelMock.object, 0))
        .returns(() => keysOverall);
      mount(<PropertyGrid
        orientation={Orientation.Vertical}
        dataProvider={dataProviderMock.object}
        selectionHandler={selectionHandlerMock.object}
        imodel={imodelMock.object}
        rulesetId={faker.random.word()} />);
      selectionHandlerMock.target.onSelect!({
        imodel: imodelMock.object,
        source: faker.random.word(),
        changeType: SelectionChangeType.Add,
        level: 0,
        keys: keysAdded,
        timestamp: new Date(),
      }, selectionProviderMock.object);
      dataProviderMock.verify((x) => x.keys = keysOverall, moq.Times.once());
    });

    it("sets data provider keys to an empty KeySet when overall selection is empty", () => {
      const emptyKeySet = new KeySet();
      const selectionProviderMock = moq.Mock.ofType<ISelectionProvider>();
      selectionProviderMock.setup((x) => x.getSelection(imodelMock.object, 0))
        .returns(() => emptyKeySet);
      mount(<PropertyGrid
        orientation={Orientation.Vertical}
        dataProvider={dataProviderMock.object}
        selectionHandler={selectionHandlerMock.object}
        imodel={imodelMock.object}
        rulesetId={faker.random.word()} />);
      selectionHandlerMock.target.onSelect!({
        imodel: imodelMock.object,
        source: faker.random.word(),
        changeType: SelectionChangeType.Clear,
        level: 0,
        keys: new KeySet(),
        timestamp: new Date(),
      }, selectionProviderMock.object);
      dataProviderMock.verify((x) => x.keys = emptyKeySet, moq.Times.once());
    });

  });

});
