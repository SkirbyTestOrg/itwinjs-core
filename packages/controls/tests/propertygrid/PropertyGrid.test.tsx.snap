// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyGrid renders correctly 1`] = `
<PropertyGrid
  dataProvider={null}
  orientation={0}
/>
`;
